### PR TITLE
Refactors high-level control flow of createWasm in preamble.js

### PR DESCRIPTION
**Pull request by lourd**
Follow-up to #11615, some minor refactoring to improve the clarity of WASM instantiation sequence, fallback, and error handling of the `createWasm` function in `preamble.js`.

### CODE REVIEW

File: src/postamble.js [hunk 1]
```diff
@@ -409,16 +409,14 @@ function exit(status, implicit) {
...
-      err(msg);
 #endif // MODULARIZE
+      err(msg);
 #else
       var msg = 'program exited (with status: ' + status + '), but noExitRuntime is set due to an async operation, so halting execution but not exiting the runtime or preventing further async execution (you can use emscripten_force_exit, if you want to force a true shutdown)';
 #if MODULARIZE
       readyPromiseReject(msg);
-#else
-      err(msg);
 #endif // MODULARIZE
```
Comment (sbc100) [inline src/postamble.js:418]: Is this change logically separate or intertwined with the preamble changes?   Might be nice to separate it if possible?

File: src/preamble.js [hunk 2]
```diff
@@ -701,14 +701,23 @@ function abort(what) {
...
-  // Throw a wasm runtime error, because a JS error might be seen as a foreign
+  // Use a wasm runtime error, because a JS error might be seen as a foreign
   // exception, which means we'd run destructors on it. We need the error to
   // simply make the program stop.
-  throw new WebAssembly.RuntimeError(what);
+  new WebAssembly.RuntimeError(what);
 #else
-  throw what;
+  what;
 #endif
```
Comment (sbc100) [inline src/preamble.js:712]: How about including `var e =` here in both arms.. seem to make the more a little more readable and 1 fewer lines.

File: tests/runner.py [hunk 3]
```diff
@@ -878,14 +878,20 @@ def assertIdentical(self, values, y, msg=None,
...
     fail_message = 'Unexpected difference:\n' + limit_size(diff)
     if not EMTEST_VERBOSE:
       fail_message += '\nFor full output run with EMTEST_VERBOSE=1.'
     if msg:
       fail_message += '\n' + msg
     self.fail(fail_message)
 
+  def assertIdenticalUrlEncoded(self, expected, actual, **kwargs):
+    """
+    URL decodes the `actual` parameter before checking for equality
```
Comment (sbc100) [inline tests/runner.py:891]: Docstring first line on the same line as the opening tripe quote.

File: tests/jsrun.py [hunk 4]
```diff
@@ -101,7 +102,7 @@ def run_js(filename, engine=None, args=[],
   if not os.path.exists(filename):
     raise Exception('JavaScript file not found: ' + filename)
 
-  command = make_command(filename, engine, args)
```
Comment (sbc100) [inline tests/jsrun.py:104]: Are any of the changes to this file used in this PR?   Can we revert this file or split out into a separate change?

File: tests/test_other.py [hunk 5]
```diff
@@ -9966,13 +9966,18 @@ def test_assertions_on_ready_promise(self):
...
     # the instance you must use .then() to get a callback with the instance.
     create_test_file('test.js', r'''
+      // Adds a noop catch handler to avoid unhandled promise rejection error for the process
+      function addCatch(promise) {
+        promise.catch(() => {});
+        return promise;
+      }
       try {
-        Module()._main;
+        addCatch(Module())._main;
```
Comment (sbc100) [inline tests/test_other.py:9975]: Does this still test for the case we are interesting it?  i.e. when the user tries to using the promise as a module?

**Comment by lourd** [comment]
Only the last two commits showing in this sequence are intended to be for this PR. All of the others are a part of #11615, which should be reviewed and merged first. If there's a way to "stack" PRs/set a diffbase in GitHub code reviews that I'm missing please let me know.

I'd also be happy to combine the two, but didn't want a bug fix to get caught up in debate over refactoring.

**Comment by sbc100** [comment]
The only way I know to "stack" changes in github is to create a PR for each change and set the "upstream"  or "target" branch in the UI to the parent..  Then as each PR lands you need to update the parent to point to master again.

**Comment by lourd** [comment]
Hm, not sure I can since the branch for the other PR is on my fork repo 🤔  Feel free to ignore this one until the other's merged.

**Pull request closed without merging**